src/System.Private.CoreLib/shared/System/BitOps.cs

```diff
@@ -5,34 +5,747 @@
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint LeadingZeroCount(uint value)
+        {
+            if (Lzcnt.IsSupported)
+            {
+                return Lzcnt.LeadingZeroCount(value);
+            }
+
+            //                                  00  01  02  2B
+            int count = (int)Log2(value); //    32  00  01  31
+            count = 31 - count; //              -1  31  30  00
+            count += IsZero(value); //          +1  +0  +0  +0
+
+            return (uint)count; //              32  31  30  00
+        }
+
+        /// <summary>
+        /// Count the number of leading zero bits in a mask.
+        /// Similar in behavior to the x86 instruction LZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint LeadingZeroCount(ulong value)
+        {
+            if (Lzcnt.X64.IsSupported)
+            {
+                return (uint)Lzcnt.X64.LeadingZeroCount(value);
+            }
+
+            // Use the 32-bit function twice.
+
+            uint hi = (uint)(value >> 32);
+            uint lo = (uint)value;
+
+            if (Lzcnt.IsSupported)
+            {
+                hi = Lzcnt.LeadingZeroCount(hi);
+                lo = Lzcnt.LeadingZeroCount(lo);
+            }
+            else
+            {
+                hi = LeadingZeroCount(hi);
+                lo = LeadingZeroCount(lo);
+            }
+
+            // Keep lo iff hi==32
+            uint m = hi & ~32u; // Zero 5th bit of hi
+            lo *= IsZero(m); // lo *= (m == 0 ? 1 : 0)
+
+            return hi + lo;
+        }
+
+        /* Legacy implementations
+        DONE https://raw.githubusercontent.com/dotnet/corefx/151a6065fa8184feb1ac4a55c89752342ab7c3bb/src/Common/src/CoreLib/System/Decimal.DecCalc.cs
+        DONE https://raw.githubusercontent.com/dotnet/corefx/62c70143cfbb08bbf03b5b8aad60c2add84a0d9e/src/Common/src/CoreLib/System/Number.BigInteger.cs
+        */
+
+        #endregion
+
+        #region TrailingZeroCount
+
+        // PR already merged: https://github.com/dotnet/coreclr/pull/22118
+
+        /// <summary>
+        /// Count the number of trailing zero bits in a mask.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(int value)
+            => TrailingZeroCount(unchecked((uint)value));
+
+        /// <summary>
+        /// Count the number of trailing zero bits in a mask.
+        /// Similar in behavior to the x86 instruction TZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint TrailingZeroCount(uint value)
```

Comment:
@tannergooding, I _think_ there is a bug here. 
[TZCNT](https://www.felixcloutier.com/x86/tzcnt) provides operand size as output when source operand is zero, ie 0->32.
The software fallback code here will return 0->0.

So unless the C# intrinsic is doing something different, the hardware branch is different to the software branch.
I have the (branchless) code to fix it, but wanted to confirm intent first.